File: 프로그래머스/3/43162. 네트워크/네트워크.py
from collections import deque
def solution(n, computers):
    # A->B 연결 ->B->C연결 일 경우 A->C도 연결이 가능함
    
    answer =0
    
    visited = [False]*(n+1)
    graph = [[] for _ in range (n+1)]
    
    for i in range (0, len(computers)) :
        start = i+1
        for j in range (0, len(computers[i])) :
            if computers[i][j] == 1 and i!=j:
                graph[i+1].append(j+1)
    
    
    def bfs (start) :
        visited[start] = True
        q = deque([start])
        while q :
            cur = q.popleft ()
            for nxt in graph[cur] :
                if not visited[nxt] :
                    visited[nxt] = True
                    q.append(nxt)
        return True
    
    for i in range (1,n+1) :
        if not visited[i] :
            if bfs(i) :
                answer +=1
    return answer
    

File: 프로그래머스/3/43162. 네트워크/네트워크.java
import java.util.*;
class Solution {
    
    int visited[];
    public int solution(int n, int[][] computers) {
   
        int[][] coms = computers;
        int answer = 0;
        int network_num =1;
        visited  = new int [n];
        
        for (int i =0; i<n; i++) {
            if (visited[i] ==0) {
                bfs(i,network_num,coms);
                network_num+=1;
                answer+=1;
            }
        }
        
        
 
        return answer;
    }
    
    private void bfs( int start,int network_num, int[][] computers) {
        ArrayDeque <Integer> q = new ArrayDeque<>();
        q.offer(start);
        visited[start] = network_num;
 
          
        while (!q.isEmpty()) {
            int cur = q.poll();
            
            for(int i =0 ;i< computers[cur].length; i++) {
                if (visited[i] == 0 && computers[cur][i] ==1) {
                    visited[i] = network_num;
                    q.offer(i);
                }
            }
            
            
        }
  
        
    } 
}